File: day1/day1.py
import numpy as np
import sys
from typing import Tuple, List

def parse(contents: bytes) -> Tuple[List[int], List[int]]:
    """
    Parses binary contents into two lists of integers. This is done without file splitting etc
    in an attempt to make the code as fast as possible.
    - Parses a number into the left list.
    - Skips spaces.
    - Parses a number into the right list.
    - Skips a newline.
    
    Args:
        contents: A bytes object containing the input data.
    
    Returns:
        A tuple of two lists: (left_list, right_list).
    """
    left_list = []
    right_list = []
    i = 0
    n = len(contents)
    
    while i < n:
        # Parse a number into the left list
        start = i
        while i < n and contents[i] >= ord('0') and contents[i] <= ord('9'):
            i += 1
        left_list.append(int(contents[start:i]))
        
        # Skip spaces
        while i < n and contents[i] == ord(' '):
            i += 1
        
        # Parse a number into the right list
        start = i
        while i < n and contents[i] >= ord('0') and contents[i] <= ord('9'):
            i += 1
        right_list.append(int(contents[start:i]))
        
        # Skip newline
        while i < n and (contents[i] == ord('\n') or contents[i] == ord('\r')):
            i += 1
    
    return left_list, right_list

def solve_part1(l1: List[int], l2: List[int]) -> int:
    '''
    Solves the first part of the problem by subtracting the two lists and taking the absolute value
    of the result.
    '''
    l1.sort()
    l2.sort()
    a1 = np.array(l1)
    a2 = np.array(l2)

    return np.sum(np.abs(a1 - a2))


def solve_part2(l1: List[int], l2: List[int]) -> int:
    # Create the counts using np.bincount
    counts = np.bincount(l2)
    similarity = 0
    
    # Iterate over l1 and calculate similarity
    for num in l1:
        if num < len(counts) and counts[num] > 0:  # Check if num is a valid "key" in counts
            similarity += num * counts[num]
    
    return similarity

if __name__ == '__main__':
    argc = len(sys.argv)
    if argc < 2:
        print('Usage: python ay1.py [input_file]')
        exit(0)
    filename = sys.argv[1]

    with open(filename, 'rb') as f:
        contents = f.read()

    l1, l2 = parse(contents)

    result1 = solve_part1(l1, l2)
    print(f'Part 1: {result1}')

    result2 = solve_part2(l1, l2)
    print(f'Part 2: {result2}')

File: day1/day1.hs
import System.Environment (getArgs)
import System.IO (withFile, IOMode(ReadMode), hGetContents)
import Control.Monad
import Data.List (sort, foldl')
import qualified Data.Map.Strict as M
import Control.DeepSeq (deepseq)


-- | Solves part 1 of the problem by summing the absolute differences of corresponding pairs
-- in two sorted lists.
solvePart1 :: ([Int], [Int]) -> Int
solvePart1 = sum . map absAndSubtractTuple . toListOfTuples . sortEach
    where
        -- | Sorts each list in a pair of lists.
        sortEach :: ([Int], [Int]) -> ([Int], [Int])
        sortEach (l1, l2) = (sort l1, sort l2)

        -- | Converts a pair of lists into a list of tuples.
        toListOfTuples :: ([a], [b]) -> [(a, b)]
        toListOfTuples = uncurry zip

        -- | Computes the absolute difference of a tuple of two numbers.
        absAndSubtractTuple :: Num a => (a, a) -> a
        absAndSubtractTuple = abs . uncurry (-)


-- | Solves part 2 by pairing each element from the left list with its count in the right list in a map
solvePart2 :: ([Int], [Int]) -> Int
solvePart2 (l1, l2) = 
    let numbers = foldl' (insertIntoFreqMap l2) M.empty l1
    in foldl' (\acc (x, y) -> acc + (x * y)) 0 (M.toList numbers)
    where
        -- | Inserts a key into a frequency map, counting occurrences of that key in the input list.
        insertIntoFreqMap :: [Int] -> M.Map Int Int -> Int -> M.Map Int Int
        insertIntoFreqMap vs m k =
            let newMap = case M.lookup k m of
                    Nothing ->  -- number not found, insert its count
                        let occurrences = length $ filter (== k) vs
                        in M.insert k occurrences m
                    Just _ -> m  -- number has already been inserted
            in newMap

-- | Reads and parses the input file into a tuple of two lists.
readAndParseFile :: FilePath -> IO ([Int], [Int])
readAndParseFile filePath = 
    withFile filePath ReadMode $ \handle -> do
        contents <- hGetContents handle
        -- `deepseq` forces an evaluation of all contents so that the file handle does not
        -- close before the contents have been read into memory.
        contents `deepseq` return (toTupleOfLists contents)
    where
        -- | Parses a string representation of pairs of integers into a tuple of two lists.
        toTupleOfLists :: String -> ([Int], [Int])
        toTupleOfLists = unzip . sublistsToTuples . map (map read) . map words . lines

        -- | Converts a list of sublists into a list of tuples, assuming each sublist has two elements.
        sublistsToTuples :: [[Int]] -> [(Int, Int)]
        sublistsToTuples = map (\[x, y] -> (x, y)) 

-- | The main function. Reads a file path from the command line arguments,
-- parses the file, and solves both parts of the problem.
main :: IO ()
main = do
    args <- getArgs
    case args of
        [filePath] -> do
            -- Parse input
            tuples <- readAndParseFile filePath
            -- Solve part 1
            let part1Result = solvePart1 tuples
            -- Solve part 2
            let part2Result = solvePart2 tuples
            -- Print results
            putStrLn $ "Part 1: " ++ show part1Result
            putStrLn $ "Part 2: " ++ show part2Result
        _ -> putStrLn "Usage: program <file-path>"
